Kata-Trainer Reference (#320)
codewars/docs: content/references/kata-trainer.md

kind: reference
---

> import ThemedImage from "@theme/ThemedImage";

Reviewer: Please change the extension of this file to `.mdx` (`kata-trainer.mdx`) since we're using React components now.

The kata trainer is the interface you're presented when you click "Train" on a kata. It is here that you write your solution, run tests and attempt to complete the kata. Since this is where you'll be spending the majority of your time, it is recommended that you familiarize yourself with the interface and its functionality. That way, you can focus entirely on solving the kata.

<ThemedImage
  alt="Kata trainer interface"
  sources={{
    light: require("./img/kata-trainer-light.png").default,
    dark: require("./img/kata-trainer-dark.png").default,
  }}

> />
> *The interface: 1) View Panel 2) Language selection 3) Language version 4) Solution Editor 5) Sample Tests Editor 6) Action buttons 7) Enable VIM/EMACS mode*

Reviewer: ```suggestion
/>

*The interface: 1) View Panel 2) Language selection 3) Language version 4) Solution Editor 5) Sample Tests Editor 6) Action buttons 7) Enable VIM/EMACS mode*
```

## View panel
On the left side of the interface you will find two views split into separate tabs:
- **Instructions**: The kata description with information about the task.
- **Output**: A view showing the test results as well as anything you choose to print to `stdout` from within your solution code.
- **Past Solutions**: If you have completed this kata before (possibly in a different language), you can view your past solutions here. If this is your first attempt at solving the kata, this tab will not be displayed.

You can switch freely between these tabs at any time during your training on the kata. Note that running tests will automatically display the Output panel.

## Editors

The right half of the interface is dedicated to writing and testing your solution. At the very top you'll find the following options:

> - **Language**: Here you select the language you wish to train in. Many kata feature translations into multiple languages, and you can train in any of them by selecting the appropriate language here. If your training setup includes multiple languages, make sure the correct one is selected before you begin writing any code.

Reviewer 2: ```suggestion
- **Language**: Here you select the language you wish to train in. Many kata feature translations into multiple languages, and you can train in any of them by selecting the appropriate language here. Make sure the correct one is selected before you begin writing any code.
```

Training routine are not the only thing which affects selected language, and I think it might be confusing to readers how it suddenly come into play here.

- **Language version**: Over time, programming languages evolve and new versions are released, which may then get added to Codewars. Select your preferred language version from the available options here. Note that your choice affects both your solution and the test suite and can have a significant impact on performance under certain circumstances.

> - **VIM/EMACS mode**: If you prefer your editor to replicate a VIM or EMACS experience, you can toggle those modes with the buttons on the right. Because the <kbd>Esc</kbd> key has a special function in VIM mode, its usual behavior of toggling full-screen mode is disabled.

Reviewer 2: ```suggestion
- **VIM/EMACS mode**: If you prefer your editor to replicate a VIM or EMACS experience, you can toggle those modes with the buttons on the right.
```

Note on 'usual behaviour' feels to me a bit out of order, because the usual behaviour were not mentioned yet. Maybe move the note about redefined Esc to the place where it's actually introduced? (see suggestion below)

Reviewer: By the way, the default is [sublime](https://codemirror.net/demo/sublime.html).

:::note
If you find that your editor no longer accepts your input or behaves in an unexpected way, make sure you haven't accidentally enabled VIM mode.
:::

### Solution
This is where you write your solution code. The solution editor features language-based syntax highlighting and auto-indentation as well as an optional VIM/EMACS mode (see above). The button in the top-right corner switches to full-screen view. If you're used to working in an IDE, you'll find the solution editor supports many of the same hotkeys for common editing tasks.

> Code within the editors is saved every time you run tests or attempt to submit your solution, allowing you to close the page and continue where you left off once you return.

Reviewer 2: Maybe it would be worth to mention that code is stored locally in the browser,  and it won't be available when training on another machine, or in another browser. Ive seen users surprised that intermediate work is not stored on the server.

Reviewer: I'm pretty sure it's only stored in the database once you test or attempt.

Kata Editor stores locally, but I don't think the trainer does. I need to check later.

### Test
This button will run your solution against the contents of the sample tests editor, offering a quick way to verify basic functionality and correctness of your code. It also ensures the contents of the solution editor are saved, should you leave the trainer and return later.

### Attempt
The attempt button will run your solution against the kata's full test suite. These test cases are not shown in the editor, but you will see their results in the Output panel. Passing all tests is a pre-requisite to submitting your solution and completing the kata. If your attempt is successful, this button will turn green and change its label to "Submit". Clicking it again will then submit your solution, marking the kata as completed, and take you to the solutions page. Any changes made to your code between these steps will require a new attempt.

## Shortcuts

The following keyboard shortcuts are available inside the kata trainer:

- <kbd>Command</kbd>+<kbd>Enter</kbd> Attempt
- <kbd>Command</kbd>+<kbd>'</kbd> Run sample tests
- <kbd>Command</kbd>+<kbd>U</kbd> Toggle focus between editors
- <kbd>Command</kbd>+<kbd>I</kbd> Cycle between Instructions and Output in the view panel

> - <kbd>Esc</kbd> Toggle full screen mode for the editor currently with focus (in VIM mode, toggles *command-line mode* on and off)

Reviewer: ```suggestion
- <kbd>Esc</kbd> Toggle full screen mode on the focused editor
```

I don't think we need to mention Vim here (or just mentioned "except Vim"). Also, it's called `Normal` mode, and doesn't toggle.
